### PR TITLE
VBM-137: Feature/add delete favorite items

### PR DESCRIPTION
**Pull request by Val-M9**
[Відео без назви - Screen Recording - 18.09.2022, 14_11_26.webm](https://user-images.githubusercontent.com/71255723/190899305-91cfadaa-afb3-4eed-82ab-0e01e6ebc838.webm)

### CODE REVIEW

File: packages/mobile/src/screens/product-info/components/price-block/styles.ts [hunk 1]
```diff
@@ -33,8 +33,8 @@ const styles = StyleSheet.create({
     borderRadius: 22,
   },
   icon: {
-    marginTop: 9,
-    marginLeft: 9.3,
+    marginTop: 7,
+    marginLeft: 7.5,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/product-info/components/price-block/styles.ts:37]: Do we really need to be that precise?

File: packages/mobile/src/screens/product-info/product-info.tsx [hunk 2]
```diff
@@ -31,33 +32,53 @@ import {
...
+      return await productApi.uploadToFavorites({ productId });
+    } catch (err) {
+      if (err instanceof HttpError) {
+        notification.error(JSON.stringify(err.message));
+      }
+    } finally {
+      dispatch(productsActions.fetchFavoritesIds());
+      setIsLoading(false);
+    }
+  };
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/product-info/product-info.tsx:80]: Put it into actions

File: packages/mobile/src/services/product/product.service.ts [hunk 3]
```diff
@@ -29,6 +37,45 @@ class ProductService {
...
   }
+
+  getFavorites(requestParams: ProductRequestDto = {}): Promise<ProductDto[]> {
+    return this.#http.load(
+      `${this.#apiPrefix}${ApiRoutes.PRODUCTS}${ProductApiRoutes.FAVORITE}`,
+      { params: requestParams },
+    );
+  }
+
+  getFavoritesIds(): Promise<Array<string> | []> {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.service.ts:48]:
```suggestion
  getFavoritesIds(): Promise<string[]> {
```

File: packages/mobile/src/services/product/product.service.ts [hunk 4]
```diff
@@ -29,6 +37,45 @@ class ProductService {
...
+    return this.#http.load(
+      `${this.#apiPrefix}${ApiRoutes.PRODUCTS}${ProductApiRoutes.FAVORITE_IDS}`,
+    );
+  }
+
+  uploadToFavorites(payload: ProductIdRequest): Promise<ProductIdRequest> {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.service.ts:54]:
```suggestion
  uploadToFavorites(payload: ProductIdRequest): Promise<ProductIdResponse> {
```

File: packages/mobile/src/services/product/product.service.ts [hunk 5]
```diff
@@ -29,6 +37,45 @@ class ProductService {
...
+        method: HttpMethod.POST,
+        contentType: HttpContentType.APPLICATION_JSON,
+        payload: JSON.stringify(payload),
+        hasAuth: true,
+      },
+    );
+  }
+
+  deleteFromFavorites(
+    payload: Record<string, unknown>,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.service.ts:67]: Why is it unknown?

File: packages/mobile/src/store/products/reducer.ts [hunk 6]
```diff
@@ -1,30 +1,69 @@
...
+  fetchFavorites,
+  fetchFavoritesIds,
+  loadProductInfo,
+} from './actions';
 
 type InitialState = {
-  products: ProductDto[] | [];
   dataStatus: DataStatus;
+  products: ProductDto[] | [];
+  current: ItemDto | undefined;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/reducer.ts:14]:
```suggestion
  current: ProductDto | null;
```

File: packages/mobile/src/store/products/reducer.ts [hunk 7]
```diff
@@ -1,30 +1,69 @@
...
+  favorites: ProductDto[] | [];
+  favoritesIds: Array<string>;
 };
 
 const initialState: InitialState = {
-  products: [],
   dataStatus: DataStatus.IDLE,
+  products: [],
+  current: undefined,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/reducer.ts:22]:
```suggestion
  current: null,
```

File: packages/mobile/src/components/svg/star/star.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { ColorPalette } from '@vse-bude/shared';
+import React from 'react';
+import Svg, { Polygon, SvgProps } from 'react-native-svg';
+
+const StarSvg = (props: SvgProps) => {
```
[the remark this answers is not in the document]
Comment (Val-M9) [reply packages/mobile/src/components/svg/star/star.tsx:5]: Turns out FeatherIcon has no fill prop as it is basically text, same as there no filled icon to swap, so I downloaded svg feather-icon

File: packages/mobile/src/screens/product-info/components/price-block/styles.ts (hunk 1 above)
Comment (Val-M9) [reply packages/mobile/src/screens/product-info/components/price-block/styles.ts:37]: actually - no, fixed

File: packages/mobile/src/services/product/product.service.ts (hunk 3 above)
Comment (Val-M9) [reply packages/mobile/src/services/product/product.service.ts:48]: fixed

File: packages/mobile/src/services/product/product.service.ts (hunk 4 above)
Comment (Val-M9) [reply packages/mobile/src/services/product/product.service.ts:54]: fixed

File: packages/mobile/src/services/product/product.service.ts (hunk 5 above)
Comment (Val-M9) [reply packages/mobile/src/services/product/product.service.ts:67]: change to string

File: packages/mobile/src/store/products/reducer.ts (hunk 6 above)
Comment (Val-M9) [reply packages/mobile/src/store/products/reducer.ts:14]: fixed and moved to product reducer

File: packages/mobile/src/store/products/reducer.ts (hunk 7 above)
Comment (Val-M9) [reply packages/mobile/src/store/products/reducer.ts:22]: fixed

File: packages/mobile/src/screens/product-info/product-info.tsx [hunk 9]
```diff
@@ -31,34 +39,70 @@ import {
...
     >();
   const id = route.params?.itemId;
+  const isFavorite = favoriteIds.includes(id);
 
   useEffect(() => {
     dispatch(productActions.loadProductInfo(id));
     dispatch(productActions.updateProductViews(id));
+    if (user) {
+      if (favoriteIds) {
+        favoriteIds.map((id) => dispatch(productsActions.addToFavorite(id)));
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/product-info/product-info.tsx:59]: ???

Pls review all this logic. It looks messy

File: packages/mobile/src/screens/product-info/product-info.tsx [hunk 10]
```diff
@@ -31,34 +39,70 @@ import {
...
-    type,
-    imageLinks,
-    views,
-    author,
-  } = product;
+  const { title, type, imageLinks, views, author } = product;
   const isAuction = type == ProductType.AUCTION;
 
+  const handleToggleFavorite = async (productId: string) => {
+    setIsLoading(true);
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/product-info/product-info.tsx:73]: You can use data status for detecting loading state

File: packages/mobile/src/screens/product-info/product-info.tsx [hunk 11]
```diff
@@ -31,34 +39,70 @@ import {
...
+    switch (Boolean(user)) {
+      case true:
+        if (isFavorite) {
+          dispatch(productsActions.deleteFromFavorite(productId))
+            .unwrap()
+            .catch((err) => notification.error(JSON.stringify(err.message)))
+            .finally(() => {
+              dispatch(productsActions.fetchFavoriteIds());
+              setIsLoading(false);
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/product-info/product-info.tsx:82]: Keep this within action

File: packages/mobile/src/services/product/product.service.ts [hunk 12]
```diff
@@ -46,6 +48,45 @@ class ProductService {
...
+    );
+  }
+
+  getFavoritesIds(): Promise<string[]> {
+    return this.#http.load(
+      `${this.#apiPrefix}${ApiRoutes.PRODUCTS}${ProductApiRoutes.FAVORITE_IDS}`,
+    );
+  }
+
+  uploadToFavorites(payload: ProductIdRequest): Promise<ProductIdRequest> {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.service.ts:65]:
```suggestion
  addToFavorites(payload: ProductIdRequest): Promise<ProductIdRequest> {
```

File: packages/mobile/src/store/products/actions.ts [hunk 13]
```diff
@@ -13,4 +17,74 @@ const loadProducts = createAsyncThunk<
...
+  (productId) => {
+    return { payload: productId };
+  },
+);
+
+const deleteFromFavoriteGuestUser = createAction<PrepareAction<string>, string>(
+  ActionType.DELETE_FROM_FAVORITE,
+  (productId) => {
+    return { payload: productId };
+  },
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/actions.ts:71]: No need to create separate actions. You may determine if the user is logged in by `userId`

File: packages/mobile/src/store/products/actions.ts [hunk 14]
```diff
@@ -13,4 +17,74 @@ const loadProducts = createAsyncThunk<
...
+);
+
+const cleanFavoriteIds = createAction<PrepareAction<[]>, string>(
+  ActionType.CLEAN_FAVORITES_IDS,
+  () => {
+    return { payload: [] };
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/actions.ts:77]: Payload is redundant in that case

**Review by AleksandrTermenzhy — CHANGES_REQUESTED** [review]

File: packages/mobile/src/services/product/product.service.ts (hunk 12 above)
Comment (Val-M9) [reply packages/mobile/src/services/product/product.service.ts:65]: fixed

File: packages/mobile/src/screens/product-info/product-info.tsx (hunk 11 above)
Comment (Val-M9) [reply packages/mobile/src/screens/product-info/product-info.tsx:82]: fixed

File: packages/mobile/src/screens/product-info/product-info.tsx (hunk 10 above)
Comment (Val-M9) [reply packages/mobile/src/screens/product-info/product-info.tsx:73]: fixed

File: packages/mobile/src/screens/product-info/product-info.tsx (hunk 9 above)
Comment (Val-M9) [reply packages/mobile/src/screens/product-info/product-info.tsx:59]: fixed, some part moved to signIn handler

File: packages/mobile/src/store/products/actions.ts (hunk 14 above)
Comment (Val-M9) [reply packages/mobile/src/store/products/actions.ts:77]: fixed

**Comment by Val-M9** [comment]
reinitialized in new branch

**Pull request closed without merging**